### ISSUE TITLE
Allow to get rid of `nix develop` "shell" logic

### ISSUE THREAD

**Issue by roberth** [NixOS/nix#7501]
**Is your feature request related to a problem? Please describe.**

We have a number of bugs and feature requests related to `nix-shell` and `nix develop`.

"Shells" appear to be built on derivations, but they are not a derivation feature. A derivation has a builder, which is a self-contained, non-interactive process that's written for a single purpose.

Currently, to create a shell, Nix has to intrude on `stdenv`'s domain and make assumptions about how the builder is implemented in order to _modify_ the derivation to support some specific code it needs to run.

This creates a situation where `nix develop` and `stdenv` are tightly coupled, which holds back both components' development and does not allow alternative derivation builders to support shells.

To break this tight coupling, I propose a new interface between `nix develop` and the expression language. Instead of taking apart a derivation, we can ask the expression nicely to produce a command that launches a shell for us.

Concretely, this means `nix develop` will perform the following pseudocode

```
nix develop () {
  if drv?devShell {
    nix run drv.devShell
  } else {
    nixDevelopLegacy(drv)
  }
}
```

By delegating the construction of a shell process to stdenv, we empower Nixpkgs to solve its own problems. This has the additional benefit that Nixpkgs allows easier collaboration, breaking changes and more risk taking in general. After all `nix` has a much stronger expectation of compatibility, because `nix` can not be pinned to projects.

Other benefits include
 - shells become first class in Nixpkgs, which is the underlying goal of #7468 ("XY problem")
 - implementing a minimal shell is trivial https://github.com/NixOS/nixpkgs/pull/132617
 - switching to a different builder shell becomes a possibility https://github.com/nixpkgs-architecture/issues/issues/7

Out of scope for this change (but relevant)
 - Producing an accurate sandbox-like environment for debugging builds. This has traditionally been a task of `nix-shell`, but it just doesn't produce a sandbox environment, leading to questions, typically: "why does this work in nix-shell, but not in my build?". This should be solved by a different command that is capable of producing an actual sandbox environment. This is a significantly different task that needs its own solution. It does also need to interface with user code, but in this case it's the derivation, not the expression that should be asked. This *slightly* special sandbox may be differentiated by _only_ the existence of a tty on stdin, or perhaps also a single special environment variable. It will be up to the builder process to read these clues and kick into a debugging mode; typically an interactive shell.

**Describe the solution you'd like**

See pseudocode above.

Similarly:

```
nix print-dev-env() {
  if drv?devShell {
    nix run drv.devShell -- --print-dev-env
  } else {
    nixPrintDevEnvLegacy(drv)
  }
}
```

**Describe alternatives you've considered**

Keep status quo, expend Nix team effort on shell features that the community can more easily fix themselves, and leave more important Nix issues and improvements behind.

**Additional context**

Solution to
 - #7468 
 - most "nix-shell" issues: https://github.com/NixOS/nix/issues?q=is%3Aopen+is%3Aissue+label%3Anix-shell

Refs
 - https://github.com/NixOS/nixpkgs/pull/206728

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by fricklerhandwerk** [comment]
Discussed in Nix team meeting 2022-12-23:

 - agreement that this is worth discussing more
 - @edolstra a `nix run`-like interface is not sufficient; would like to make the expression export a bash script and/or plain environment variables
 - @edolstra: multi-shell support is actually possible by providing aliases that call into bash
   - @roberth: this is a never ending pit of feature requests, and a feature/tool that's beneficial to users outside the Nix ecosystem, so it should be its own project that Nixpkgs can use to implement this _other shell_ feature

**Comment by tomberek** [comment]
`develop` does a few things in order:
1) analyze the derivation and create a build environment for the derivation
2) initialize the build environment using the "--rcfile" trick in order to go last, juggle some env vars for convenience, and call hooks

Most of the non-`nix debug` usages care about (2), and we've built tooling to sidestep or undo the effects of (1), eg: mkShell, devshell, and so forth. My interpretation of this proposal is to make it easier and less of a hack to bypass (1).

**Comment by tomberek** [comment]
## Steps ahead
- [x] idea is approved
- [ ] need `nix develop` and `nix print-dev-env` interface/specification descriptions
- [ ] implementation, POC -> MVP -> experimental -> complete

## Random notes:
- `nix debug`
- breakpointHook - cntr
- nix bundle --bundle builderInSandbox
- nix bundle --bundle builderInOCI
- nix bundle --bundle in QEMU

- nix develop --with nixpkgs#make-me-a-devShell
- nix develop --with custom#pkgs.defaultDevShell
- 
- allow for overriding of `get-env.sh`
    - passthru.devShell
    - passthru.devShell { drv = builtins.getDerivationEnvironment drv; get-env = "source $something"; }
    - meta.get-env?
    - passthru.devShell = self // { builder = "source $my-own-stdenv";};
    - passthru.devShell = mkShell { name = "asdfas"; builder = "something-custom";}
    - passthru.devShell = drv: function-that-creates-JSON

**Comment by fricklerhandwerk** [comment]
Related: https://github.com/NixOS/nix/pull/4702#issuecomment-1614561883

**Comment by roberth** [comment]
@fricklerhandwerk, in https://github.com/NixOS/nix/pull/4702#issuecomment-1614561883 you propose to

> reduce the Nix command line interface to essentially `nix run`, `nix build`, `nix eval`, and `nix store`

Did you mean that list to be exhaustive - ie discourage even a simple `nix develop` entrypoint, or did you omit it not to balloon the scope there?

I believe this extra entrypoint - `nix develop` - is extremely valuable as a "thin waist" around which Nix-aware tooling can interoperate with Nix-aware projects.

**Comment by fricklerhandwerk** [comment]
Not exactly exhaustive, but a good approximation. I haven't made up my mind on `nix develop`: It is centered around `stdenv` and the idea that the build environment is somehow run by a shell. That's was done historically for good practical reasons, but it's not a hard requirement. Of course Nix could have `nix develop` start *any* (even the current) shell with the `builder`'s environment as defined by the store derivation, but how useful would that really be? A lot of `stdenv` happens within the build script. Arguably the whole notion of a shell environment for derivation builds is native to `stdenv`.

**Comment by lisanna-dettwyler** [comment]
I think I'm going to take a shot at finishing the implementation of this

**Comment by lisanna-dettwyler** [comment]
@roberth a couple things that would be good to get your input on:
- The `hello.devShell` design seems to leave room for only one kind of shell (bash), is there a way we can adapt the design to allow for devshells to be defined in the future that use different shells under the hood?
- I think the biggest lift with this work is going to be testing it, to make sure the new implementation exactly mimics the old behavior, especially with regards to things like rc files in the user's environment. Any ideas you have for how we could approach the testing would be appreciated.

**Comment by roberth** [comment]
> * allow for devshells to be defined in the future that use different shells under the hood?

Yes, it also has an env file that should work in any shell, and the implementation of `devShell` could support dispatch on `SHELL` for instance.

> Any ideas you have for how we could approach the testing would be appreciated.

The bulk of it will be in `devShellTools` tests, see e.g. this [WIP](https://github.com/NixOS/nixpkgs/pull/330822/changes#diff-30b8b5f32c0ba0d71e65138a990aefbce146e8ca0444742285f4ed483f97dc3e).
The Nix side of things becomes a lot simpler. We keep the existing tests for the current/legacy behaviors, and add simple unit tests to exercise the simple "exec" code path.
